### PR TITLE
[SE-4399] feat: add LTI URL copy feature

### PR DESCRIPTION
**Pull request by gabor-boros**
This PR adds a feature to copy a block's LTI launch URL to easily integrate with LTI consumers. The copy button reaches out to Studio to the URL instead of composing it on the front-end to give flexibility in future refactors if necessary.

**Dependencies**: 

* https://github.com/edx/edx-platform/pull/27712
* https://github.com/edx/edx-platform/pull/27411

**Screenshots**: 

https://user-images.githubusercontent.com/19173947/119609778-6e991f80-bdf8-11eb-84c3-732684018540.mov

**Sandbox URL**: https://library-authoring.pr27712.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

* Login to studio
* Switch to the libraries tab
* Create a new library
* Create a new block in the library
* Click on Copy LTI Url button
* Validate the URL is copied
* Check the success message appeared

**Author notes and concerns**:

N/A

**Reviewers**
- [ ] @jvdm
- [ ] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @gabor-boros! I've created [OSPR-5799](https://openedx.atlassian.net/browse/OSPR-5799) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by natabene** [comment]
@gabor-boros Thank you for your contribution. Please let me know once it is ready for our review.

**Review by jvdm — APPROVED** [review]
:+1:

- [x] I tested this: (describe what you tested)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation
- [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

@gabor-boros Approving, but I do have one small comment regarding accessibility, which I leave to your discretion to address or not:

We don't have a "copy confirmation" feedback for customers, do we? Can we add a message somewhere stating "Copied: http://localhost/..."?

**Comment by gabor-boros** [comment]
@natabene Thank you! I'll ping you here when we are ready 😊 

@jvdm  Thank you for the feedback! I just added some visual feedback for the user in a form of a success message. Could you please check now? Also, the video is updated.

**Comment by jvdm** [comment]
> @natabene Thank you! I'll ping you here when we are ready blush
> 
> @jvdm Thank you for the feedback! I just added some visual feedback for the user in a form of a success message. Could you please check now? Also, the video is updated.

Amazing! Thanks for improving things so quickly.

**Review by jvdm — APPROVED** [review]
:+1:

- [x] I tested this: (describe what you tested)
- [x] I read through the code
- [x] I checked for accessibility issues
- ~[ ] Includes documentation~
- ~[ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~

**Comment by gabor-boros** [comment]
@natabene This PR is ready for edX review.

**Comment by gabor-boros** [comment]
@natabene  this is now ready for edX review! Thank you for the patience.

**Comment by pomegranited** [comment]
@arbrandes could you review this as core committer?

@natabene says [TNL are ready to delegate this to core committers](https://openedx.slack.com/archives/C015KC8CN76/p1628534531018500), and you're the only on with merge rights on this repo :)

**Comment by arbrandes** [comment]
@pomegranited, sure!

@gabor-boros, the code looks good and we've tested it extensively.  But I think we missed something along the way: there needs to be a way to hide this button when LTI is not in use.  Ideally, we'd use [the model](https://github.com/edx/edx-platform/pull/27411/files#diff-36022deef8607c7a4647c8f2620b4d9ed283d5b41077e966bfd097585e0ebe7cR121) that defines whether a particular library is configured to be used by an LTI consumer: if it is configured, this button shows.  If not, it doesn't.  Is that something you can look into before we merge this?

**Comment by gabor-boros** [comment]
@arbrandes oh! That was totally missed, I'm so sorry about that!

I'll look into that tomorrow morning to be able to merge as soon as we can!

**Comment by gabor-boros** [comment]
@arbrandes I looked into this. As I can see we could ​expose the `lti_tool`'s data (or info about existence) as part of the `ContentLibraryMetadataSerializer`.

The `ContentLibraryMetadataSerializer` could be extended with an `lti_enabled` `BooleanField` field that would be true if the `lti_tool` on the model is not `None`, otherwise it would return `False`. Before jumping on the implementation, how would that sound to you?

**Comment by arbrandes** [comment]
@gabor-boros, that certainly works for me.  Good idea.  Unless @jvdm can think of an objection, you can go ahead and do it.

**Comment by jvdm** [comment]
> @gabor-boros, that certainly works for me. Good idea. Unless @jvdm can think of an objection, you can go ahead and do it.

It looks like a good direction moving forward. But, heads up @gabor-boros: We will change the relationship between Libraries and LTI configuration (aka. `lti_tool`) from 1:N to M:N. Let's keep this in mind. I think just making sure the model encapsulates the `is_lti_enable` logic in a method or property. Then, let the serializer pull the boolean out of it. That would simplify refactoring it later.

**Comment by gabor-boros** [comment]
@jvdm to not step on each other's feet, would you recommend to wait for the M:N refactoring or may I go ahead and implement the is_lti_tool_enabled method as soon as I get there?

**Comment by jvdm** [comment]
@gabor-boros:

> @jvdm to not step on each other's feet, would you recommend to wait for the M:N refactoring or may I go ahead and implement the is_lti_tool_enabled method as soon as I get there?

I'll probably only have the refactoring done later today. So, if you don't want to wait until then, please move forward. If you encapsulate the "enabled" attribute, it will be easy to change it without breaking the logic sitting on top of it.

**Comment by jvdm** [comment]
@gabor-boros: The refactoring is ready for review, I've added [a property to the library model](https://github.com/edx/edx-platform/pull/27411/files#diff-36022deef8607c7a4647c8f2620b4d9ed283d5b41077e966bfd097585e0ebe7cR144) which I think is what we need here.

**Comment by arbrandes** [comment]
@jvdm, thanks for coordinating with @gabor-boros.  I think the `allow_lti` property will do what we need nicely.

**Comment by gabor-boros** [comment]
That is awesome @jvdm, thank you for adding that! I'll go ahead and extend the serializer on the backend and parsed data on the front-end.

**Comment by gabor-boros** [comment]
There are some issues with my devstack I was not able to fix to check the changes. I'll try to fix that as soon as I can and refactor the PR.

**Pull request merged**
